File: packages/node/src/sdk/masking.ts
/**
 * Sensitive-data masking utilities for DT3 Commons log events.
 *
 * The masking engine traverses nested objects and arrays, redacts values whose
 * field names are sensitive, and returns a copy so callers' source data is
 * never changed.
 */

export const DEFAULT_SENSITIVE_FIELDS = [
  'password',
  'passwd',
  'pwd',
  'secret',
  'token',
  'access_token',
  'refresh_token',
  'authorization',
  'api_key',
  'apikey',
  'private_key',
  'credit_card',
  'card_number',
  'ssn',
  'nic',
  'national_id',
  'email',
  'phone',
] as const;

/**
 * Configuration for a {@link MaskingEngine}.
 */
export interface MaskingOptions {
  /** Additional field names to redact, matched case-insensitively. */
  sensitiveFields?: Iterable<string>;
  /** The value substituted for sensitive values. Defaults to `[REDACTED]`. */
  replacementValue?: string;
  /** Whether `mask` includes paths of fields that were redacted. */
  trackMaskedFields?: boolean;
  /** Whether masking is active. Defaults to `true`. */
  enabled?: boolean;
}

/**
 * Result returned from a masking operation.
 */
export interface MaskingResult<T> {
  /** A recursively copied value with sensitive field values redacted. */
  data: T;
  /** Paths of redacted fields, or an empty array when tracking is disabled. */
  maskedFields: string[];
}

/**
 * Recursively masks sensitive field values in data structures.
 */
export class MaskingEngine {
  private readonly sensitiveFields: ReadonlySet<string>;
  private readonly replacementValue: string;
  private readonly trackMaskedFields: boolean;
  private readonly enabled: boolean;

  /**
   * Create a masking engine using default and caller-supplied sensitive fields.
   *
   * @param options - Masking behavior and additional sensitive field names.
   */
  constructor(options: MaskingOptions = {}) {
    const fields = new Set<string>();

    for (const field of DEFAULT_SENSITIVE_FIELDS) {
      fields.add(field.toLocaleLowerCase());
    }

    for (const field of options.sensitiveFields ?? []) {
      if (typeof field === 'string') {
        fields.add(field.toLocaleLowerCase());
      }
    }

    this.sensitiveFields = fields;
    this.replacementValue = options.replacementValue ?? '[REDACTED]';
    this.trackMaskedFields = options.trackMaskedFields ?? false;
    this.enabled = options.enabled ?? true;
  }

  // PUBLIC_INTERFACE
  /**
   * Return a recursively copied, masked version of the supplied data.
   *
   * @typeParam T - The input data type.
   * @param data - Object, array, primitive, or nested combination to process.
   * @returns The copied data and optional paths of masked fields.
   */
  public mask<T>(data: T): MaskingResult<T> {
    if (!this.enabled) {
      return {
        data: this.copyValue(data) as T,
        maskedFields: [],
      };
    }

    const result = this.maskValue(data, '');
    return {
      data: result.data as T,
      maskedFields: this.trackMaskedFields ? result.maskedFields : [],
    };
  }

  private maskValue(value: unknown, path: string): MaskingResult<unknown> {
    if (Array.isArray(value)) {
      const maskedFields: string[] = [];
      const data = value.map((item, index) => {
        const result = this.maskValue(item, `${path}[${index}]`);
        maskedFields.push(...result.maskedFields);
        return result.data;
      });

      return { data, maskedFields };
    }

    if (this.isPlainObject(value)) {
      const maskedFields: string[] = [];
      const data: Record<string, unknown> = {};

      for (const [key, childValue] of Object.entries(value)) {
        const childPath = path ? `${path}.${key}` : key;

        if (this.sensitiveFields.has(key.toLocaleLowerCase())) {
          data[key] = this.replacementValue;
          maskedFields.push(childPath);
          continue;
        }

        const result = this.maskValue(childValue, childPath);
        data[key] = result.data;
        maskedFields.push(...result.maskedFields);
      }

      return { data, maskedFields };
    }

    return { data: this.copyValue(value), maskedFields: [] };
  }

  private copyValue(value: unknown): unknown {
    if (Array.isArray(value)) {
      return value.map((item) => this.copyValue(item));
    }

    if (this.isPlainObject(value)) {
      return Object.fromEntries(
        Object.entries(value).map(([key, childValue]) => [key, this.copyValue(childValue)]),
      );
    }

    return value;
  }

  private isPlainObject(value: unknown): value is Record<string, unknown> {
    return value !== null && typeof value === 'object' && !Array.isArray(value);
  }
}

File: packages/python/dt3_sdk/masking.py
"""Sensitive-data masking utilities for DT3 Commons log events."""

from __future__ import annotations

from copy import deepcopy
from typing import Any, Iterable, List, Mapping, Optional, Set, Tuple

DEFAULT_SENSITIVE_FIELDS = (
    "password",
    "passwd",
    "pwd",
    "secret",
    "token",
    "access_token",
    "refresh_token",
    "authorization",
    "api_key",
    "apikey",
    "private_key",
    "credit_card",
    "card_number",
    "ssn",
    "nic",
    "national_id",
    "email",
    "phone",
    "prompt",
    "response",
    "kavia.prompt",
    "kavia.response",
)


class MaskingEngine:
    """Recursively masks sensitive values without mutating source data."""

    def __init__(
        self,
        sensitive_fields: Optional[Iterable[str]] = None,
        replacement_value: str = "[REDACTED]",
        track_masked_fields: bool = False,
        enabled: bool = True,
    ) -> None:
        """Create a masking engine with configurable masking rules.

        Args:
            sensitive_fields: Additional field names to mask, matched case-insensitively.
            replacement_value: Value substituted for sensitive field values.
            track_masked_fields: Whether masked field paths are returned by ``mask``.
            enabled: Whether masking is active.
        """
        configured_fields = sensitive_fields or ()
        self._sensitive_fields: Set[str] = {
            field.casefold()
            for field in (*DEFAULT_SENSITIVE_FIELDS, *configured_fields)
            if isinstance(field, str)
        }
        self.replacement_value = replacement_value
        self.track_masked_fields = track_masked_fields
        self.enabled = enabled

    # PUBLIC_INTERFACE
    def mask(self, data: Any) -> Tuple[Any, List[str]]:
        """Return a masked copy of data and, when configured, masked field paths.

        Args:
            data: A mapping, list, primitive, or nested combination to process.

        Returns:
            A tuple containing a deep-copied masked value and masked field paths.
            The paths list is empty unless ``track_masked_fields`` is enabled.
        """
        if not self.enabled:
            return deepcopy(data), []

        masked_data, masked_fields = self._mask_value(data, path="")
        return masked_data, masked_fields if self.track_masked_fields else []

    def _mask_value(self, value: Any, path: str) -> Tuple[Any, List[str]]:
        """Recursively process one value while preserving its container structure."""
        if isinstance(value, Mapping):
            masked_mapping = {}
            masked_fields: List[str] = []

            for key, child_value in value.items():
                key_text = str(key)
                child_path = f"{path}.{key_text}" if path else key_text

                if key_text.casefold() in self._sensitive_fields:
                    masked_mapping[key] = self.replacement_value
                    masked_fields.append(child_path)
                    continue

                masked_child, child_fields = self._mask_value(child_value, child_path)
                masked_mapping[key] = masked_child
                masked_fields.extend(child_fields)

            return masked_mapping, masked_fields

        if isinstance(value, list):
            masked_list = []
            masked_fields = []

            for index, child_value in enumerate(value):
                child_path = f"{path}[{index}]"
                masked_child, child_fields = self._mask_value(child_value, child_path)
                masked_list.append(masked_child)
                masked_fields.extend(child_fields)

            return masked_list, masked_fields

        if isinstance(value, tuple):
            masked_items = []
            masked_fields = []

            for index, child_value in enumerate(value):
                child_path = f"{path}[{index}]"
                masked_child, child_fields = self._mask_value(child_value, child_path)
                masked_items.append(masked_child)
                masked_fields.extend(child_fields)

            return tuple(masked_items), masked_fields

        return deepcopy(value), []
